### ISSUE TITLE
Parallelize scans for shape-full expressions

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#12707]
As a Chapel programmer, I'm happy that `+ scan A` is parallelized for default and block-distributed arrays, but disappointed to find that `+ scan A: int` (for example), is not.  I'd like to extend the same parallel implementation that exists for these arrays to simple expressions / promotions whose result domain is of one of the supported array types.

**Comment by bradcray** [comment]
I was thinking that using Vass's support for querying the shapes of iterator records I could do this easily by creating the result array before calling the array's scan routine, passing it in, and essentially reusing the existing implementation.  But what I forgot is that we don't support random access to an iterator record, so can't index into `(A: int)` the way we can for `A`.  But I'm thinking probably we should support this?  (i.e., make such promoted cases less like general iterator expressions and more like a pseudo-array?).  I think the next step is to propose an issue for this.  This seems like it would also support the longstanding desire to have a promoted field access like `A.x` act like an array of x.type.

[edit: Opened https://github.com/chapel-lang/chapel/issues/28926 to capture this idea]

**Comment by psath** [comment]
+1 for wanting this feature

I have a userspace jagged array (call it "outer" with lots of "inners") and it would be nice to be able to parallelize
`+ scan [i in 0...#outer.size] outer[i].inner.size`
i.e. to construct a serialized `offsets` array

**Comment by bradcray** [comment]
@psath : In your code, is it the case that it's working, but just using a serial, rather than parallel, scan at present?

Would a potential short-term workaround be:

```chapel
var sizes = [i in 0..#outer.size] outer[i].inner.size
const offsets = + scan sizes;
```

This has the obvious overhead of using a temporary array…  But our scans use temporary arrays as well, so this may not be the end of the world.

Also, are you representing your jagged arrays using an array of records that store an array field?

Thanks!

**Comment by psath** [comment]
> [@psath](https://github.com/psath) : In your code, is it the case that it's working, but just using a serial, rather than parallel, scan at present?

Oh the code works great today w/ Chapel 2.8! I only ended up here from the compiler warning that it is being serialized. I was honestly stoked to be able to do that in a one-liner 😄 but it's functionally correct for the up to 100ks of elements I've tried.

Once I scale up to i in the 10s of billions, having it parallelized will matter more though.

> Also, are you representing your jagged arrays using an array of records that store an array field?

Edit: yes. If this is an anti-pattern or there is a more chapel-tastic way, I'm all ears. Still in prototyping so lots of latitude to reorganize as needed.

**Comment by bradcray** [comment]
> If this is an anti-pattern or there is a more chapel-tastic way, I'm all ears.

No, I think that's the best practice today.  Just wanted to make sure I had the right mental picture.  Here's a sample code capturing the pattern I think you've got, Paul, as well as some stylistic rewrites (that don't help with parallelization) and the workaround proposed above (that does, but requires a temp array):  [ATO](https://ato.pxeger.com/run?1=rVMxTsNAECzo_AeklWgSQZzQoYQUPACB0lBEKRZ7E6847tDeOQEiXkKTBh7Fa1ifnQBSggBR-bw7OzM39j6_ZgXekVmtXsow7Zy87e0LZU5yGMEyAZijgOdH6gPbMGgKbC1JH8a9ND2tmhMQQjNInpKkarsykMAQxqxIqEDHvQlYWsCoVcGH3B4kC-FAxrYiWN-TJXS7cCE8Y4sG5iSenW0E3XTqKXjlPASfof2gPojzae0insc8SaPBWKwsb6RqlnY0GuUuUdAYMrBwcoPiSpt_unOlV9N_5dvqKA58r3buhNR2TvcdS2VQac0twpUOQqGP9fVb12UAH9gY8CSMpv19ErXN3N0i2z_kcFU4Qx0UwYd_sLQts93aZ9apjuzSPQItCga2M3D6V8Rm4xTjYChUk9WARsuZfrVfhOc24WlqP7Fcb0mzLOuleQc)

Meanwhile, I'm trying to remember whether this issue was an easy lift or a more difficult one.

**Comment by psath** [comment]
Ahh, thanks for the variants! That captures the current prototype, and are good examples for me to keep teaching my C-brain the shorter Array Expressions way.

**Comment by bradcray** [comment]
@psath : I've got a 5-line patch that gets us to a parallel solution by inserting a temporary array to capture the iteratable expression within the implementation of the scan.  This is like the workaround above, simply within the scan implementation rather than at the user level.  How off-putting is the creation of that temporary array for your use-case?  (i.e., can you afford another `outer.size`-element array?

Refreshing my memory on the code and my [2019 comment above](https://github.com/chapel-lang/chapel/issues/12707#issuecomment-479604446), I think doing an array-less implementation would be a significant undertaking, requiring us to change the representation of loop expressions in the implementation.  I've opened https://github.com/chapel-lang/chapel/issues/28926 to capture that concept.

**Comment by psath** [comment]
Yeah I think we can afford the temporary. Willing to trade memory for speed

**Comment by bradcray** [comment]
@psath : This should be implemented on `main` now, so should appear in this month's 2.9 release.  If you get a chance to try it out before then, I'll be curious whether you have any negative experiences that should be addressed before the release goes out (if not, there's always 2.10 :) ).
